packages/SwingSet/src/kernel/kernel.js

```diff
@@ -283,24 +283,41 @@ export default function buildKernel(kernelEndowments, kernelOptions = {}) {
     }
   }
 
-  function removeVatManager(vatID, reason) {
-    const old = ephemeral.vats.get(vatID);
-    ephemeral.vats.delete(vatID);
-    const err = reason ? Error(reason) : undefined;
-    old.notifyTermination(err); // XXX TODO: most of the places where notifyTermination gets passed don't need it
-    return old.manager.shutdown();
+  function rejectOldPromises(vatID, reason) {
+    const promisesToReject = kernelKeeper.cleanupAfterTerminatedVat(vatID);
+    const error = reason ? makeError(reason) : VAT_TERMINATION_ERROR;
+    for (const kpid of promisesToReject) {
+      resolveToError(kpid, error, vatID);
+    }
   }
 
-  function terminateVat(vatID, reason) {
+  function terminateVatOnFailure(vatID, reason) {
```

Comment:
`terminateVat` and `terminateVatOnFailure` are essentially identical (modulo variable name rotation) but for the notification call.  Exposing separate entry points on the public API (i.e., the device API) is fine, but implementations should be unified so they don't drift apart under maintenance.  The way `sendTerminationNotification` was done here is a good path to follow.

Reply:
I worked on this for a while before. With renewed purpose, I found a way to turn the non-shared part into a lambda to pass along.